### ISSUE TITLE
Amplify owner:ID using @auth does not register

### ISSUE THREAD

**Issue by NiksanJP** [aws-amplify/amplify-js#10212]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React Native

### Amplify APIs

GraphQL API, DataStore

### Amplify Categories

auth, api

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

When using @auth(rules: [{ allow: owner, provider: userPools, ownerField: "owner" }]), even without the auth the owner field does not get initialized for some models. 

Schema:
![image](https://user-images.githubusercontent.com/40626274/184829074-0e116706-b085-4ef9-ace2-9b94a3afbe56.png)

There is not much difference in Message and Item

API results
![image](https://user-images.githubusercontent.com/40626274/184829122-d29c1719-957a-4139-9da5-8ef7d8f3572c.png)

owner is getting populated for Message but not for Item

### Expected behavior

owner should be populated for all tables.

### Reproduction steps

Run the schema provided

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### aws-exports.js

/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "ap-southeast-2",
    "aws_cognito_identity_pool_id": "XXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "aws_cognito_region": "ap-southeast-2",
    "aws_user_pools_id": "XXXXXXXXXXXXXXXXXXXXXX",
    "aws_user_pools_web_client_id": "XXXXXXXXXXXXX",
    "oauth": {},
    "aws_cognito_username_attributes": [
        "EMAIL"
    ],
    "aws_cognito_social_providers": [],
    "aws_cognito_signup_attributes": [
        "EMAIL",
        "ADDRESS",
        "BIRTHDATE",
        "NAME",
        "PHONE_NUMBER",
        "PREFERRED_USERNAME"
    ],
    "aws_cognito_mfa_configuration": "OFF",
    "aws_cognito_mfa_types": [
        "SMS"
    ],
    "aws_cognito_password_protection_settings": {
        "passwordPolicyMinLength": 8,
        "passwordPolicyCharacters": [
            "REQUIRES_LOWERCASE",
            "REQUIRES_NUMBERS",
            "REQUIRES_UPPERCASE"
        ]
    },
    "aws_cognito_verification_mechanisms": [
        "EMAIL"
    ],
    "aws_user_files_s3_bucket": "XXXXXXX-storage-devXXXXXXXXXXXXXXXXdev",
    "aws_user_files_s3_bucket_region": "ap-southeast-2",
    "aws_appsync_graphqlEndpoint": "http://XXXXXXXXXXXXX/graphql",
    "aws_appsync_region": "ap-southeast-2",
    "aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS",
    "aws_appsync_apiKey": "XXXXXXXXXXXXXXXXXX",
    "aws_appsync_dangerously_connect_to_http_endpoint_for_testing": true
};

export default awsmobile;

**Comment by chrisbonifacio** [comment]
Hi! Thanks for raising this issue. If you're using Cognito User Pools and naming your owner field "owner", you can remove the `provider: userPools` and `ownerField: "owner"` part of your schema. So, all you need is `{ allow: owner }`.

Try making this change to your schema, then run `amplify push` and let me know if that fixes the issue for you.

**Comment by NiksanJP** [comment]
> 

even when running it separately it does not work at all.

**Comment by NiksanJP** [comment]
Attempt 1:
![image](https://user-images.githubusercontent.com/40626274/184873456-49b218d5-b31c-4f3a-935d-d37c0a27d3ed.png)

![image](https://user-images.githubusercontent.com/40626274/184873522-f70d89ff-8c8a-468b-aafb-5659331f9a1a.png)


Attempt 2:
![image](https://user-images.githubusercontent.com/40626274/184873582-00191ff1-20e6-4192-9715-be86ecf48570.png)
![image](https://user-images.githubusercontent.com/40626274/184873696-253d09c7-3e19-4b62-a334-8b1a533afb9d.png)

So basically I cannot use some words....
How was this production ready?

**Comment by NiksanJP** [comment]
There are instances where I cannot even run mock API due to table names, any docos for table name reqs?

**Comment by NiksanJP** [comment]
Another update, so the above schema works with mock API but does not work with real server.
I am just tryna find the best way to get my app launch that has been stuck due to these errors.

I am just going to use Message(n) as my table names because there are no docos at all for names.
![image](https://user-images.githubusercontent.com/40626274/184909303-c160f31d-e1af-4037-b660-c48773c8ac08.png)

This is the only solution ATM

**Comment by chrisbonifacio** [comment]
I think the issue might be that you are using two owner fields on your models.
Do you have DataStore enabled and intend to use it in your application?

**Comment by NiksanJP** [comment]
> I think the issue might be that you are using two owner fields on your models. Do you have DataStore enabled and intend to use it in your application?

Yes I am using two owner fields, can I not do that?
I do have DataStore enabled on my react native app.

**Comment by chrisbonifacio** [comment]
I think you will run into issues with DataStore when you have more than one `owner` type auth rules that have READ access on a model, there is an open issue on the android repo about this: https://github.com/aws-amplify/amplify-android/issues/1728

**Comment by NiksanJP** [comment]
> I think you will run into issues with DataStore when you have more than one `owner` type auth rule that has READ access on a model, there is an open issue on the android repo about this: [aws-amplify/amplify-android#1728](https://github.com/aws-amplify/amplify-android/issues/1728)

SO there is no fix on this ?
So for all amplify-based applications, we can just act as another user and run API calls?

There is a work around this but not through Datastore which is custom resolvers....
We cannot use custom queries and mutations with DataStore can we?

**Comment by chrisbonifacio** [comment]
> we can just act as another user and run API calls?
Not sure what you mean by this, with an owner auth rule only authenticated User Pool users can make API calls but you can't pretend to be another user. 

Only the creator of a record has permission to manipulate it unless you grant access to other users explicitly. The API is deny-by-default so everything is restricted so as you add auth rules you are gradually increasing access/privileges to your users.

The issue in DataStore is that there cannot be two owners of a record that both have READ access, that's all. There is a fix already in the works. 

> We cannot use custom queries and mutations with DataStore can we?
You can but you'd have to trigger them via graphql directly. When using DataStore you are using the DataStore API which, for React Native, interfaces with either AsyncStorage or SQLite. The graphql queries and mutations occur as a background process to sync records from the client to the server and vice versa.

It sounds like, for your use case, you should use a non-DataStore API and model your Message similar to this:

```graphql
type Message
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "sender", operations: [create, read] }
      {
        allow: owner
        ownerField: "recipient"
        operations: [create, read]
      }
    ]
  ) {
  id: ID!
  content: String!
  received: Boolean!
  senderID: ID! @index(name: "bySender", sortKeyFields: ["content"])
  sender: User @belongsTo(fields: ["senderID"])
  recipientID: ID!
    @index(name: "byRecipient", sortKeyFields: ["content"])
  recipient: User @belongsTo(fields: ["recipientID"])
}
```

**Comment by NiksanJP** [comment]
Is there a function that I can use so who ever creates the item, the username gets on the field automatically.
Like @function("getcognitoattributes-{env}").
I think I read that somewhere but did not give any description.

**Comment by chrisbonifacio** [comment]
>   this does not stop me from acting as another user and calling an API

Can you explain? Do you mean via an XSS attack or some other method?

> Is there a function that I can use so who ever creates the item, the username gets on the field automatically.

So, there are a couple ways the owner can be set. Yes, you could use a custom mutation with a lambda resolver to set any ownerField you define or the generic `owner` one by Amplify. 

something similar to this

```graphql
type Mutation {
  sendMessage(data: Message): Message @function(name: "sendMessageFunction-${env}")
}
```

you can add the function that resolves this mutation with `amplify add function`

The `@function` directive can also be combined with `@auth` to further restrict who has access to invoke the lambda.

For more information on this: https://docs.amplify.aws/cli/graphql/custom-business-logic/#create-a-custom-query-or-mutation

For it to be automatic, the rule has to be `{ allow: owner }` and Amplify will populate the owner field for you with the information of the authenticated User Pool user that created the record.

> Behind the scenes, Amplify will automatically add a owner: String field to each record which contains the record owner's identity information upon record creation.
> 
> By default, the Cognito user pool's user information is populated into the owner field. The value saved includes sub and username in the format <sub>::<username>. The API will authorize against the full value of <sub>::<username> or sub / username separately and return username. You can alternatively configure [OpenID Connect as an authorization provider](https://docs.amplify.aws/cli/graphql/authorization-rules/#using-oidc-authorization-provider).

For more info on owner based auth: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access

**Comment by chrisbonifacio** [comment]
As a side note, I tried reproducing the response you got earlier with a non-DataStore enabled API. Same schema and combined `createMessage` and `createItem` mutations but the `receiverusername` was populated for both.

<img width="1164" alt="Screenshot 2022-08-16 at 2 03 42 PM" src="https://user-images.githubusercontent.com/16952986/184948258-ba2fb87d-3288-4331-b9d7-5fc7e8a70a97.png">

Here you can see that users `chris` and `mary` can see messages between each other.

logged in as `chris`
<img width="801" alt="Screenshot 2022-08-16 at 2 08 23 PM" src="https://user-images.githubusercontent.com/16952986/184949372-9fe07e60-3d27-4ef5-8461-d8021f165d11.png">

logged in as `mary`
<img width="801" alt="Screenshot 2022-08-16 at 2 08 07 PM" src="https://user-images.githubusercontent.com/16952986/184949433-edf4e992-df15-4cba-949b-e64698b95701.png">

but a third user, `jude`, who is not defined as an owner of the Message record shown above does not receive the messages between `chris` and `mary`
<img width="801" alt="Screenshot 2022-08-16 at 2 08 41 PM" src="https://user-images.githubusercontent.com/16952986/184949708-c3539f98-6696-44db-b9ac-1a310becf14e.png">

So, perhaps the issue is with the mock api capabilities because when testing the queries and mutations via AppSync and normal usage, they work as expected.

**Comment by chrisbonifacio** [comment]
Actually, I just tried with mock api and got a response where both mutations had `owner` populated...

<img width="1081" alt="Screenshot 2022-08-16 at 2 40 31 PM" src="https://user-images.githubusercontent.com/16952986/184954570-ac866a22-4d19-4917-8bf6-cccf89bf3b42.png">

What version of the Amplify CLI (`@aws-amplify/cli`) do you have installed?

**Comment by NiksanJP** [comment]
@chrisbonifacio Thank you for the explanation.
From my side, I just wanted to say that while using DataStore and not mock API.
I cannot populate the owner field automatically.
I was able to populate both owner fields in mock and Appsync.

The alternate solution that EVERYONE is using for DataStore is manually inputting the owner field.
I think this is a flaw and the user can take advantage by using other people's username instead of theirs.

I can confirm that there is a walk around using functions and custom resolvers but while using them,
I find it very hard to use the custom resolvers's subscription which is very buggy.

**Comment by NiksanJP** [comment]
Just wondering if this will cause any trouble with API?

**Comment by NiksanJP** [comment]
Another find, I managed to Add only ONE ownerfield to a single table, once I add it to another table (making it 2 ownerfields in n tables) it shows "Data Store subscriptionError, connection failed {"error": "message":...."

**Comment by chrisbonifacio** [comment]
> Another find, I managed to Add only ONE ownerfield to a single table, once I add it to another table (making it 2 ownerfields in n tables) it shows "Data Store subscriptionError, connection failed {"error": "message":...."

what is the full message returned from that subscription error?

> I cannot populate the owner field automatically.
I was able to populate both owner fields in mock and AppSync

Can you share the code snippets where you are trying to create a record with DataStore? Are you seeing any errors logged to the console when you try this?

>The alternate solution that EVERYONE is using for DataStore is manually inputting the owner field.

Can you share what that workaround looks like? I'm not sure I'm familiar.

>I think this is a flaw and the user can take advantage by using other people's username instead of theirs.

Are you saying that a user can change the value of `owner` on a record that exists on the server and belongs to another user via DataStore?

**Comment by chrisbonifacio** [comment]
Hi 👋 Closing this as we have not heard back from you. If you are still experiencing this issue and in need of assistance, please feel free to comment and provide us with any information previously requested by our team members so we can re-open this issue and be better able to assist you.

Thank you!
